File: web/run-manager/src/shared/api/contract/saveGames.ts
// web/run-manager/src/shared/api/contract/saveGames.ts
import { z } from "zod";

import {
  rendererSchema,
  type WatchDevice,
  type WatchRenderer,
  watchDeviceSchema,
} from "@/shared/api/contract/enums";

export const saveGameStatusSchema = z.enum(["created", "running", "paused", "finished", "failed"]);

export const saveUnlockInspectionStatusSchema = z.enum(["not_inspected", "inspected"]);

export const saveUnlockTargetStatusSchema = z.enum([
  "pending",
  "locked",
  "succeeded",
  "failed",
  "skipped",
]);

export const saveAttemptStatusSchema = z.enum(["running", "succeeded", "failed"]);

export const savePolicyArtifactSchema = z.enum(["latest", "best"]);

export const policyPlaybackModeSchema = z.enum(["deterministic", "stochastic"]);

export const managedSaveCourseSetupSchema = z.object({
  id: z.string(),
  save_game_id: z.string(),
  difficulty: z.string().nullable(),
  cup_id: z.string().nullable(),
  course_id: z.string().nullable(),
  policy_run_id: z.string(),
  policy_artifact: savePolicyArtifactSchema,
  engine_setting_raw_value: z.number().int().min(0).max(100),
  created_at: z.string(),
  updated_at: z.string(),
});

export const managedSaveCupSetupSchema = z.object({
  id: z.string(),
  save_game_id: z.string(),
  difficulty: z.string().nullable(),
  cup_id: z.string(),
  vehicle_id: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const managedSaveUnlockTargetSchema = z.object({
  sequence_index: z.number(),
  kind: z.string(),
  status: saveUnlockTargetStatusSchema,
  label: z.string(),
  difficulty: z.string().nullable(),
  cup_id: z.string().nullable(),
  course_id: z.string().nullable(),
});

export const managedSaveUnlockProgressSchema = z.object({
  inspection_status: saveUnlockInspectionStatusSchema,
  completed_count: z.number(),
  total_count: z.number(),
  unlocked_vehicle_count: z.number(),
  unlocked_vehicle_ids: z.array(z.string()),
  next_target: managedSaveUnlockTargetSchema.nullable(),
  targets: z.array(managedSaveUnlockTargetSchema),
});

export const managedSaveAttemptSchema = z.object({
  id: z.string(),
  save_game_id: z.string(),
  target_kind: z.string().nullable(),
  status: saveAttemptStatusSchema,
  difficulty: z.string().nullable(),
  cup_id: z.string().nullable(),
  course_id: z.string().nullable(),
  started_at: z.string(),
  finished_at: z.string().nullable(),
  finish_position: z.number().nullable(),
  finish_time_s: z.number().nullable(),
  failure_reason: z.string().nullable(),
});

export const saveGameRunnerSettingsSchema = z.object({
  device: watchDeviceSchema,
  renderer: rendererSchema,
  policy_mode: policyPlaybackModeSchema,
  attempt_seed: z.number().int().min(0).max(4_294_967_295).nullable(),
  recording_enabled: z.boolean(),
  recording_input_hud_enabled: z.boolean(),
  recording_upscale_factor: z.number().int().min(1).max(4),
  recording_path: z.string().nullable(),
  target_restart_on_retire: z.boolean().default(false),
  target_clear_goal: z.number().int().min(0).default(1),
  keep_failed_recordings: z.boolean().default(false),
});

export const managedSaveGameSchema = z.object({
  id: z.string(),
  name: z.string(),
  status: saveGameStatusSchema,
  runner_active: z.boolean(),
  save_path: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
  last_finished_at: z.string().nullable(),
  runner_settings: saveGameRunnerSettingsSchema,
  unlock_progress: managedSaveUnlockProgressSchema.nullable(),
  attempts: z.array(managedSaveAttemptSchema),
  course_setups: z.array(managedSaveCourseSetupSchema),
  cup_setups: z.array(managedSaveCupSetupSchema),
});

export const managedSaveGameStatusSchema = z.object({
  id: z.string(),
  name: z.string(),
  status: saveGameStatusSchema,
  runner_active: z.boolean(),
  save_path: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
  last_finished_at: z.string().nullable(),
  runner_settings: saveGameRunnerSettingsSchema,
  unlock_progress: managedSaveUnlockProgressSchema.nullable(),
});

export const saveGamesResponseSchema = z.object({
  save_games: z.array(managedSaveGameSchema),
});

export const saveGameStatusResponseSchema = z.object({
  save_game: managedSaveGameStatusSchema,
});

export const createSaveGameResponseSchema = z.object({
  save_game: managedSaveGameSchema,
});

export const updateSaveGameRunnerSettingsResponseSchema = z.object({
  save_game: managedSaveGameSchema,
});

export const upsertSaveCourseSetupResponseSchema = z.object({
  save_game: managedSaveGameSchema,
});

export const saveEngineTuningCourseSetupRecommendationSchema = z.object({
  difficulty: z.string().nullable(),
  cup_id: z.string(),
  course_id: z.string(),
  vehicle_id: z.string(),
  engine_setting_raw_value: z.number().int().min(0).max(100),
  mean_score: z.number().nullable(),
  finish_count: z.number().int().nonnegative(),
});

export const importSaveEngineTuningResponseSchema = z.object({
  recommendations: z.array(saveEngineTuningCourseSetupRecommendationSchema),
});

export const deleteSaveGameResponseSchema = z.object({
  deleted: z.boolean(),
});

export const openSaveGameDirectoryResponseSchema = z.object({
  opened: z.boolean(),
});

export type ManagedSaveGame = z.infer<typeof managedSaveGameSchema>;
export type ManagedSaveGameStatus = z.infer<typeof managedSaveGameStatusSchema>;
export type ManagedSaveAttempt = z.infer<typeof managedSaveAttemptSchema>;
export type ManagedSaveCourseSetup = z.infer<typeof managedSaveCourseSetupSchema>;
export type ManagedSaveCupSetup = z.infer<typeof managedSaveCupSetupSchema>;
export type ManagedSaveUnlockProgress = z.infer<typeof managedSaveUnlockProgressSchema>;
export type ManagedSaveUnlockTarget = z.infer<typeof managedSaveUnlockTargetSchema>;
export type PolicyPlaybackMode = z.infer<typeof policyPlaybackModeSchema>;
export type SaveGameRunnerSettings = z.infer<typeof saveGameRunnerSettingsSchema>;
export type SaveEngineTuningCourseSetupRecommendation = z.infer<
  typeof saveEngineTuningCourseSetupRecommendationSchema
>;
export type SaveAttemptStatus = z.infer<typeof saveAttemptStatusSchema>;
export type SavePolicyArtifact = z.infer<typeof savePolicyArtifactSchema>;
export type SaveUnlockTargetStatus = z.infer<typeof saveUnlockTargetStatusSchema>;

export interface CareerModeRunnerLaunchRequest {
  attemptSeed: string | null;
  device: WatchDevice;
  policyMode: PolicyPlaybackMode;
  recordingEnabled: boolean;
  recordingInputHudEnabled: boolean;
  recordingUpscaleFactor: number;
  recordingPath: string | null;
  renderer: WatchRenderer | null;
  saveGameId: string;
  singleTarget: boolean;
  perfectRun: boolean;
  keepFailedRecordings: boolean;
  targetClearGoal: number;
  target: ManagedSaveUnlockTarget | null;
}

export interface SaveGameRunnerSettingsUpdateRequest {
  attemptSeed: string | null;
  device: WatchDevice;
  policyMode: PolicyPlaybackMode;
  recordingEnabled: boolean;
  recordingInputHudEnabled: boolean;
  recordingUpscaleFactor: number;
  recordingPath: string | null;
  renderer: WatchRenderer;
  saveGameId: string;
  targetRestartOnRetire: boolean;
  targetClearGoal: number;
  keepFailedRecordings: boolean;
}

File: src/rl_fzerox/apps/run_manager/api/payloads/save_games.py
# src/rl_fzerox/apps/run_manager/api/payloads/save_games.py
from __future__ import annotations

from rl_fzerox.core.career_mode.execution.context import SaveAttemptExecutionContext
from rl_fzerox.core.manager import (
    ManagedSaveAttempt,
    ManagedSaveCourseSetup,
    ManagedSaveCupSetup,
    ManagedSaveGame,
    ManagedSaveUnlockProgress,
    ManagedSaveUnlockTarget,
)


def save_course_setup_payload(
    assignment: ManagedSaveCourseSetup,
) -> dict[str, object]:
    return {
        "id": assignment.id,
        "save_game_id": assignment.save_game_id,
        "difficulty": assignment.difficulty,
        "cup_id": assignment.cup_id,
        "course_id": assignment.course_id,
        "policy_source_kind": assignment.policy_source_kind,
        "policy_source_id": assignment.policy_source_id,
        "policy_artifact": assignment.policy_artifact,
        "engine_setting_raw_value": assignment.engine_setting_raw_value,
        "created_at": assignment.created_at,
        "updated_at": assignment.updated_at,
    }


def save_cup_setup_payload(
    assignment: ManagedSaveCupSetup,
) -> dict[str, object]:
    return {
        "id": assignment.id,
        "save_game_id": assignment.save_game_id,
        "difficulty": assignment.difficulty,
        "cup_id": assignment.cup_id,
        "vehicle_id": assignment.vehicle_id,
        "created_at": assignment.created_at,
        "updated_at": assignment.updated_at,
    }


def save_attempt_payload(attempt: ManagedSaveAttempt) -> dict[str, object]:
    return {
        "id": attempt.id,
        "save_game_id": attempt.save_game_id,
        "target_kind": attempt.target_kind,
        "status": attempt.status,
        "difficulty": attempt.difficulty,
        "cup_id": attempt.cup_id,
        "course_id": attempt.course_id,
        "started_at": attempt.started_at,
        "finished_at": attempt.finished_at,
        "finish_position": attempt.finish_position,
        "finish_time_s": attempt.finish_time_s,
        "failure_reason": attempt.failure_reason,
    }


def save_game_runner_settings_payload(save_game: ManagedSaveGame) -> dict[str, object]:
    return {
        "device": save_game.runner_device,
        "renderer": save_game.runner_renderer,
        "policy_mode": save_game.runner_policy_mode,
        "attempt_seed": save_game.runner_attempt_seed,
        "recording_enabled": save_game.runner_recording_enabled,
        "recording_input_hud_enabled": save_game.runner_recording_input_hud_enabled,
        "recording_upscale_factor": save_game.runner_recording_upscale_factor,
        "recording_path": None
        if save_game.runner_recording_path is None
        else str(save_game.runner_recording_path),
        "target_restart_on_retire": save_game.runner_target_restart_on_retire,
        "target_clear_goal": save_game.runner_target_clear_goal,
        "keep_failed_recordings": save_game.runner_keep_failed_recordings,
        "reload_policy_between_attempts": save_game.runner_reload_policy_between_attempts,
    }


def save_attempt_execution_context_payload(
    context: SaveAttemptExecutionContext,
) -> dict[str, object]:
    """Serialize one resolved career-runner attempt context."""

    return {
        "save_game": save_game_payload(context.save_game),
        "attempt": save_attempt_payload(context.attempt),
        "target": {
            "kind": context.target.kind,
            "label": context.target.label,
            "difficulty": context.target.difficulty,
            "cup_id": context.target.cup_id,
            "course_id": context.target.course_id,
        },
        "course_setup": save_course_setup_payload(context.course_setup),
        "cup_setup": save_cup_setup_payload(context.cup_setup),
        "policy_source": {
            "kind": context.policy_source.kind,
            "id": context.policy_source.id,
            "name": context.policy_source.name,
            "mutable": context.policy_source.mutable,
            "source_run_id": context.policy_source.source_run_id,
            "source_run_name": context.policy_source.source_run_name,
        },
        "policy_artifact": context.policy_artifact,
        "policy_path": str(context.policy_path),
    }


def save_unlock_target_payload(target: ManagedSaveUnlockTarget) -> dict[str, object]:
    return {
        "sequence_index": target.sequence_index,
        "kind": target.kind,
        "status": target.status,
        "label": target.label,
        "difficulty": target.difficulty,
        "cup_id": target.cup_id,
        "course_id": target.course_id,
    }


def save_unlock_progress_payload(progress: ManagedSaveUnlockProgress) -> dict[str, object]:
    return {
        "inspection_status": progress.inspection_status,
        "completed_count": progress.completed_count,
        "total_count": progress.total_count,
        "unlocked_vehicle_count": progress.unlocked_vehicle_count,
        "unlocked_vehicle_ids": list(progress.unlocked_vehicle_ids),
        "next_target": None
        if progress.next_target is None
        else save_unlock_target_payload(progress.next_target),
        "targets": [save_unlock_target_payload(target) for target in progress.targets],
    }


def save_game_payload(
    save_game: ManagedSaveGame,
    *,
    runner_active: bool = False,
    unlock_progress: ManagedSaveUnlockProgress | None = None,
    attempts: tuple[ManagedSaveAttempt, ...] = (),
    course_setups: tuple[ManagedSaveCourseSetup, ...] = (),
    cup_setups: tuple[ManagedSaveCupSetup, ...] = (),
) -> dict[str, object]:
    return {
        "id": save_game.id,
        "name": save_game.name,
        "status": save_game.status,
        "save_path": str(save_game.save_path),
        "created_at": save_game.created_at,
        "updated_at": save_game.updated_at,
        "last_finished_at": save_game.last_finished_at,
        "runner_active": runner_active,
        "runner_settings": save_game_runner_settings_payload(save_game),
        "unlock_progress": None
        if unlock_progress is None
        else save_unlock_progress_payload(unlock_progress),
        "attempts": [save_attempt_payload(attempt) for attempt in attempts],
        "course_setups": [save_course_setup_payload(assignment) for assignment in course_setups],
        "cup_setups": [save_cup_setup_payload(assignment) for assignment in cup_setups],
    }


def save_game_status_payload(
    save_game: ManagedSaveGame,
    *,
    runner_active: bool = False,
    unlock_progress: ManagedSaveUnlockProgress | None = None,
) -> dict[str, object]:
    return {
        "id": save_game.id,
        "name": save_game.name,
        "status": save_game.status,
        "save_path": str(save_game.save_path),
        "created_at": save_game.created_at,
        "updated_at": save_game.updated_at,
        "last_finished_at": save_game.last_finished_at,
        "runner_active": runner_active,
        "runner_settings": save_game_runner_settings_payload(save_game),
        "unlock_progress": None
        if unlock_progress is None
        else save_unlock_progress_payload(unlock_progress),
    }
